rfcs/FragmentArguments.md

```diff
@@ -0,0 +1,378 @@
+
+The other scoping options are:
+
+- Global, i.e. a fragment argument is just syntactic sugar for an operation
+  variable.
+  - This is what we implemented at Meta, and it was terrible for all the reasons
+    you can think of.
+- Recursively local, i.e. the variable takes on any parent fragment argument
+  definition
+  - This preserves the concept of the value being some sort of recursively
+    scoped variable.
+  - However, as explained above, keeping track of what's happening, and
+    preventing fragment conflicts, becomes really difficult.
+
+We're choosing to explicitly _allow_ overriding operation variables, as the
+local scope means you can clearly see whether a variable is scoped to the
+fragment or operation.
+
+## New Validation Rule: Fragment Argument Definitions Used in Fragment
+
+With local scope, this rule is very simple.
+
+```
+fragment Foo($x: Int) on User {
+  name
+}
+```
+
+would be invalid.
+
+Additionally,
+
+```
+fragment Foo($x: Int!) on User {
+  ...Bar
+}
+
+fragment Bar {
+  number(x: $x)
+}
+```
+
+would also be invalid: even though `$x` is used underneath Foo, it is used
+outside of Foo's explicit definition. In this context, `$x` in Bar is actually
+an operation variable.
+
+However, this would be valid:
+
+```
+fragment Foo($x: Int!) on User {
+  ...Bar(x: $x)
+}
+
+fragment Bar($x: Int) {
+  number(x: $x)
+}
+```
+
+### Consideration: how strict should this rule be?
+
+As an initial RFC, I'd advocate for encouraging the _strictest_ version of this
+rule possible: any argument defined on a fragment must be explicitly used by
+that same fragment. It would be easy to relax the rule later, but very difficult
+to do the reverse.
+
+It's clearly more composable if, when changing a child fragment, you don't need
+to worry about modifying argument definitions on parent fragments callsites.
+However, we could in the future allow annotating argument definitions with
+`@deprecated`.
+
+## Updated Validation Rule: Required Arguments are Provided
+
+We update
+[Required Arguments](https://spec.graphql.org/draft/#sec-Required-Arguments) to
+include fragment spreads. This makes the validation's first two bullets:
+
+- For each Field, **Fragment Spread** or Directive in the document.
+- Let _arguments_ be the set of argument definitions of that Field, **Fragment**
+  or Directive.
+
+With this rule, the below example is invalid, even if the argument
+`User.number(x:)` is nullable in the schema.
+
+```
+fragment Foo on User {
+  ...Bar
+}
+
+fragment Bar($x: Int!) on User {
+  number(x: $x)
+}
+```
+
+### Potential Alternative: Default Value indicates _required or not_, and `!` indicates _non-null or nullable_.
+
+If we were writing the language from scratch, I'd advocate for making _all_
+argument definitions without a default value to be required, regardless of their
+nullability. If you want to make a nullable argument optional, you do so by
+adding a `= null` to its definition.
```

Comment:
That doesn't work really well either. If you do

```
fragment Bar($x: Int = null) { number(x: $x) }
```

then what would `$x` be in `...Bar`? Would it be `null` or `undefined`? To be consistent with other default values, I'd expect it to be `null` but then there's no way to pass `undefined` (no value).

Not sure what a good solution would be.  Maybe add a `undefined` keyword and allow a subset of union types...

```
fragment Bar($x: Int | undefined)
```

Reply:
I think allowing undefined variables was a mistake :)

However, the reason this version is labelled as an *alternative* is because it would try to fix that mistake in one place (fragment arguments) and not elsewhere (operation variables, field/directive arguments, input objects). At the very least we'd want to unify with either operation variables or (field/directive arguments and input objects). Since fixing the mistake of undefined inputs is outside the scope of this feature, I'm opting to keep the same behavior we already have (namely, unset gets resolved to either the fragment defined argument default value, or, if there is no default and the variable is nullable, resolves to null or the field defined argument default value).

Reply:
> I think allowing undefined variables was a mistake :)

Gotcha 👍 I think I agree. I cannot think of a use case where a undefined variable or field argument is used (but would be interested to learn about them if any).

For the sake of consistency, I'd rather keep the same behaviour as operation variables though. It took me enough time to understand how it was working that I don't want to learn something new for fragments 😅